File: scripts/update_frontmatter_descriptions.py
#!/usr/bin/env python3
"""
Обновляет frontmatter во всех content/*.md:
- удаляет priority, targetAudience, readingTime, tags
- задаёт description из текста статьи (заменяет плейсхолдер «Краткое описание» и пустое)
"""
from __future__ import annotations

import re
import sys
from pathlib import Path

try:
    import yaml
except ImportError:
    print("Need PyYAML: pip install pyyaml", file=sys.stderr)
    sys.exit(1)

REMOVE_KEYS = frozenset({"priority", "targetAudience", "readingTime", "tags"})
PLACEHOLDERS = frozenset(
    {
        "краткое описание",
        "краткое описание.",
        "todo",
        "tbd",
        "",
    }
)

ROOT = Path(__file__).resolve().parents[1]
CONTENT = ROOT / "content"


def split_frontmatter(raw: str) -> tuple[str | None, str]:
    if not raw.startswith("---"):
        return None, raw
    # First line is ---
    lines = raw.split("\n")
    if lines[0].strip() != "---":
        return None, raw
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            fm = "\n".join(lines[1:i])
            body = "\n".join(lines[i + 1 :])
            if body.startswith("\n"):
                body = body[1:]
            elif not body:
                pass
            return fm, body
    return None, raw


def strip_markdown_prose(s: str) -> str:
    s = re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", s)
    s = re.sub(r"!\[([^\]]*)\]\([^)]+\)", "", s)
    s = re.sub(r"[`*_#]+", "", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def remove_fenced_code(text: str) -> str:
    out = []
    i = 0
    n = len(text)
    while i < n:
        if text.startswith("```", i):
            end = text.find("```", i + 3)
            if end == -1:
                break
            i = end + 3
            while i < n and text[i] in "\r\n":
                i += 1
            continue
        out.append(text[i])
        i += 1
    return "".join(out)


def first_summary_sentence(body: str, title: str, max_len: int = 280) -> str:
    body = remove_fenced_code(body)
    # убрать первую строку если это одинокий заголовок H1 дублирующий title
    paragraphs = re.split(r"\n\s*\n", body)
    collected = []
    for block in paragraphs:
        block = block.strip()
        if not block:
            continue
        lines = [ln.strip() for ln in block.split("\n") if ln.strip()]
        if not lines:
            continue
        # пропуск чистых заголовков
        if all(ln.startswith("#") for ln in lines):
            continue
        # таблицы — пропуск
        if "|" in block and block.count("|") >= 4:
            continue
        prose_lines = []
        for ln in lines:
            if ln.startswith("#"):
                continue
            if ln.startswith("|") and "---" in ln:
                continue
            prose_lines.append(ln)
        if not prose_lines:
            continue
        chunk = " ".join(prose_lines)
        # списки: взять первые пункты как краткий текст
        if chunk.startswith("- ") or re.match(r"^\d+\.\s", chunk):
            chunk = strip_markdown_prose(chunk)
            # ограничить длину списка — первые 2 пункта
            parts = re.split(r"\s*-\s+", chunk)
            if len(parts) > 2:
                chunk = parts[0] + ": " + " ".join(parts[1:3]) if parts[0] else " ".join(parts[:3])
        else:
            chunk = strip_markdown_prose(chunk)
        if len(chunk) < 25:
            continue
        collected.append(chunk)
        if len(" ".join(collected)) > 80:
            break
    text = " ".join(collected).strip()
    if not text:
        text = strip_markdown_prose(title) if title else ""
    # предложения
    text = re.sub(r"\s+", " ", text)
    # обрезка по предложениям
    out = []
    total = 0
    for part in re.split(r"(?<=[.!?])\s+", text):
        if not part:
            continue
        if total + len(part) > max_len and out:
            break
        out.append(part)
        total += len(part) + 1
        if total >= max_len * 0.6 and len(out) >= 2:
            break
    result = " ".join(out).strip()
    if len(result) > max_len:
        result = result[: max_len - 1].rsplit(" ", 1)[0] + "…"
    return result or (strip_markdown_prose(title) if title else "Материал базы знаний.")


def process_file(path: Path) -> bool:
    raw = path.read_text(encoding="utf-8")
    fm_raw, body = split_frontmatter(raw)
    if fm_raw is None:
        return False
    try:
        data = yaml.safe_load(fm_raw) or {}
    except yaml.YAMLError:
        print(f"YAML error: {path}", file=sys.stderr)
        return False
    if not isinstance(data, dict):
        return False

    changed = False
    for k in REMOVE_KEYS:
        if k in data:
            del data[k]
            changed = True

    title = data.get("title") or ""
    desc = data.get("description")
    desc_str = (desc if isinstance(desc, str) else "") or ""
    need_desc = (
        not desc_str.strip()
        or desc_str.strip().lower() in PLACEHOLDERS
        or desc_str.strip() == "Краткое описание"
    )
    if need_desc:
        new_d = first_summary_sentence(body, str(title))
        data["description"] = new_d
        changed = True
    elif any(k in fm_raw for k in REMOVE_KEYS):
        # ключи уже удалили из dict — перезапишем файл
        changed = True

    if not changed:
        return False

    # Стабильный порядок ключей для читаемости
    preferred = ["title", "weight", "draft", "description"]
    ordered: dict = {}
    for k in preferred:
        if k in data:
            ordered[k] = data[k]
    for k, v in data.items():
        if k not in ordered:
            ordered[k] = v

    dump = yaml.dump(
        ordered,
        allow_unicode=True,
        default_flow_style=False,
        sort_keys=False,
        width=1000,
    )
    new_raw = "---\n" + dump + "---\n" + body
    if new_raw != raw:
        path.write_text(new_raw, encoding="utf-8")
        return True
    return False


def main() -> None:
    n = 0
    for path in sorted(CONTENT.rglob("*.md")):
        if process_file(path):
            n += 1
            print(path.relative_to(ROOT))
    print(f"Updated {n} files", file=sys.stderr)


if __name__ == "__main__":
    main()

File: scripts/update_frontmatter_descriptions.mjs
/**
 * Обновляет frontmatter во всех content/*.md:
 * - удаляет priority, targetAudience, readingTime, tags
 * - задаёт description из текста статьи (заменяет «Краткое описание» и пустое)
 */
import { readFileSync, writeFileSync } from "fs";
import { readdirSync, statSync } from "fs";
import { dirname, join, relative } from "path";
import { fileURLToPath } from "url";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const CONTENT = join(ROOT, "content");

const REMOVE_LINE_PREFIXES = [
  /^priority:\s*/,
  /^targetAudience:\s*/,
  /^readingTime:\s*/,
];

const PLACEHOLDERS = new Set([
  "краткое описание",
  "краткое описание.",
  "todo",
  "tbd",
  "",
]);

function walk(dir, out = []) {
  for (const name of readdirSync(dir)) {
    const p = join(dir, name);
    if (statSync(p).isDirectory()) walk(p, out);
    else if (name.endsWith(".md")) out.push(p);
  }
  return out;
}

function splitFrontmatter(raw) {
  if (!raw.startsWith("---\n") && !raw.startsWith("---\r\n")) return null;
  const rest = raw.slice(4);
  const m = /\r?\n---\r?\n/.exec(rest);
  if (!m) return null;
  const fm = rest.slice(0, m.index);
  const body = rest.slice(m.index + m[0].length);
  return { fm, body };
}

function removeFencedCode(text) {
  let out = "";
  let i = 0;
  while (i < text.length) {
    if (text.startsWith("```", i)) {
      const end = text.indexOf("```", i + 3);
      if (end === -1) break;
      i = end + 3;
      while (i < text.length && /[\r\n]/.test(text[i])) i++;
      continue;
    }
    out += text[i];
    i++;
  }
  return out;
}

function stripMd(s) {
  return s
    .replace(/\[([^\]]+)\]\([^)]+\)/g, "$1")
    .replace(/!\[([^\]]*)\]\([^)]+\)/g, "")
    .replace(/[`*_#]+/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

function firstSummary(body, title, maxLen = 280) {
  body = removeFencedCode(body);
  const paragraphs = body.split(/\n\s*\n/);
  const collected = [];
  for (const block of paragraphs) {
    const b = block.trim();
    if (!b) continue;
    const lines = b.split("\n").map((l) => l.trim()).filter(Boolean);
    if (!lines.length) continue;
    if (lines.every((l) => l.startsWith("#"))) continue;
    if (b.includes("|") && (b.match(/\|/g) || []).length >= 4) continue;
    const proseLines = [];
    for (const ln of lines) {
      if (ln.startsWith("#")) continue;
      if (/\|/.test(ln) && /---/.test(ln)) continue;
      proseLines.push(ln);
    }
    if (!proseLines.length) continue;
    let chunk = proseLines.join(" ");
    if (/^(- |\d+\.\s)/.test(chunk)) {
      chunk = stripMd(chunk);
      const parts = chunk.split(/\s*-\s+/);
      if (parts.length > 2) {
        chunk = parts[0] + ": " + parts.slice(1, 3).join(" ");
      }
    } else {
      chunk = stripMd(chunk);
    }
    if (chunk.length < 25) continue;
    collected.push(chunk);
    if (collected.join(" ").length > 80) break;
  }
  let text = collected.join(" ").trim();
  if (!text) text = stripMd(title || "");
  text = text.replace(/\s+/g, " ");
  const sentences = text.split(/(?<=[.!?])\s+/).filter(Boolean);
  let out = [];
  let total = 0;
  for (const part of sentences) {
    if (!part) continue;
    if (total + part.length > maxLen && out.length) break;
    out.push(part);
    total += part.length + 1;
    if (total >= maxLen * 0.6 && out.length >= 2) break;
  }
  let result = out.join(" ").trim();
  if (result.length > maxLen) {
    const cut = result.slice(0, maxLen - 1).lastIndexOf(" ");
    result = (cut > 40 ? result.slice(0, cut) : result.slice(0, maxLen - 1)) + "…";
  }
  return result || (stripMd(title) || "Материал базы знаний.");
}

/** Удаляет из массива строк FM блок tags и строки priority/targetAudience/readingTime */
function filterFmLines(lines) {
  const out = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (REMOVE_LINE_PREFIXES.some((re) => re.test(line))) continue;
    if (/^tags:\s*$/.test(line.trim())) {
      while (i + 1 < lines.length && /^\s+-\s/.test(lines[i + 1])) i++;
      continue;
    }
    if (/^tags:\s+\S/.test(line.trim())) {
      // tags: [a, b] одной строкой
      continue;
    }
    out.push(line);
  }
  return out;
}

function parseDescriptionValue(line) {
  const m = /^description:\s*(.*)$/.exec(line);
  if (!m) return null;
  let v = m[1].trim();
  if (
    (v.startsWith('"') && v.endsWith('"')) ||
    (v.startsWith("'") && v.endsWith("'"))
  ) {
    v = v.slice(1, -1);
  }
  return v;
}

function needsNewDescription(val) {
  const s = (val || "").trim().toLowerCase();
  return PLACEHOLDERS.has(s) || s === "краткое описание";
}

/** description: | или > — не трогаем */
function isMultilineDescriptionStart(line) {
  return /^\s*description:\s*([|>]|-)\s*$/.test(line);
}

function processFile(path) {
  const raw = readFileSync(path, "utf8");
  const sp = splitFrontmatter(raw);
  if (!sp) return false;

  let { fm, body } = sp;
  /** Hugo shortcodes в description ломают YAML — не трогаем такие файлы */
  if (/description:\s*\{\{/.test(fm)) return false;
  const lines = fm.split(/\r?\n/);
  const filtered = filterFmLines(lines);

  let title = "";
  let descLineIdx = -1;
  let descVal = "";
  let skipReplace = false;

  for (let i = 0; i < filtered.length; i++) {
    const line = filtered[i];
    const tm = /^title:\s*(.*)$/.exec(line);
    if (tm) {
      let t = tm[1].trim();
      if (
        (t.startsWith('"') && t.endsWith('"')) ||
        (t.startsWith("'") && t.endsWith("'"))
      )
        t = t.slice(1, -1);
      title = t;
    }
    if (/^\s*description:\s*/.test(line)) {
      descLineIdx = i;
      if (isMultilineDescriptionStart(line)) {
        skipReplace = true;
        break;
      }
      descVal = parseDescriptionValue(line) ?? "";
    }
  }

  if (skipReplace) {
    // только удаление ключей
    const newFm = filtered.join("\n");
    const newRaw = `---\n${newFm}\n---\n${body}`;
    if (newRaw !== raw) {
      writeFileSync(path, newRaw, "utf8");
      return true;
    }
    return false;
  }

  const needDesc = needsNewDescription(descVal);
  const newDesc = needDesc ? firstSummary(body, title) : descVal;

  if (descLineIdx >= 0 && needDesc) {
    filtered[descLineIdx] = `description: ${yamlEscapeScalar(newDesc)}`;
  } else if (needDesc && descLineIdx < 0) {
    // вставить description после draft если есть, иначе после title
    let ins = 0;
    for (let i = 0; i < filtered.length; i++) {
      if (/^draft:\s*/.test(filtered[i])) {
        ins = i + 1;
        break;
      }
      if (/^title:\s*/.test(filtered[i])) ins = i + 1;
    }
    filtered.splice(ins, 0, `description: ${yamlEscapeScalar(newDesc)}`);
  }

  const newFm = filtered.join("\n");
  const newRaw = `---\n${newFm}\n---\n${body}`;
  if (newRaw === raw) return false;
  writeFileSync(path, newRaw, "utf8");
  return true;
}

function yamlEscapeScalar(s) {
  if (/[:#\n\r\t]/.test(s) || s.startsWith('"') || s.startsWith("'")) {
    return (
      '"' +
      s.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n") +
      '"'
    );
  }
  return s;
}

let n = 0;
for (const p of walk(CONTENT)) {
  try {
    if (processFile(p)) {
      n++;
      console.log(relative(ROOT, p));
    }
  } catch (e) {
    console.error(p, e);
  }
}
console.error(`Updated ${n} files`);
